### PR TITLE
FEATURE: add Date.formatCldr helper

### PR DESCRIPTION
**Pull request by litovchenko1**
**What I did**

This change adds CLDR formatting capabillity to Date Eel helper,
to match functionality of format.date Fluid ViewHelper.

**How to verify it**

Just use `${Date.formatCldr(someDateObject, 'dd MMMM yyyy HH:mm:ss', 'de')}` to test. Also test different combinations with missing arguments.

**Checklist**

- [x] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [x] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)

### CODE REVIEW

File: Neos.Eel/Classes/Helper/DateHelper.php [hunk 1]
```diff
@@ -57,6 +71,31 @@ public function format($date, $format)
     }
 
     /**
+     * Format a date to a string with a given cldr format
+     *
+     * @param \DateTime $date
+     * @param string $cldrFormat Format string in CLDR format (see http://cldr.unicode.org/translation/date-time)
+     * @param string $locale String locale - example (de|en|ru_RU). Must be one of Neos\Flow\I18n\Cldr\Reader\DatesReader::FORMAT_LENGTH_*'s constants.
```
Comment (albe) [inline Neos.Eel/Classes/Helper/DateHelper.php]: I think the second part of that line is a bit off, it's not a format length parameter :)

File: Neos.Eel/Classes/Helper/DateHelper.php [hunk 2]
```diff
@@ -57,6 +71,31 @@ public function format($date, $format)
...
+     */
+    public function formatCldr($date = null, $cldrFormat = null, $locale = null)
+    {
+        if ($date === null) {
+            $date = new \DateTime('now');
+        }
+        if ($cldrFormat === null) {
+            $cldrFormat = 'dd MMMM yyyy HH:mm:ss';
+        }
+        if ($locale === null) {
```
Comment (aertmann) [inline Neos.Eel/Classes/Helper/DateHelper.php]: probably a good idea to not be strict about non set values to allow `${Date.formatCldr(date, '', 'fr')}` as well as `${Date.formatCldr(date, null, 'fr')}`

Comment (albe) [inline Neos.Eel/Classes/Helper/DateHelper.php]: Good point, bu then please be explicitly unstrict ;) i.e. `$cldrFormat === null || $cldrFormat === ''`
Relying on the PHP type juggling behaviour is a fallacy

**Comment by kitsunet** [comment]
I would actually simply enforce giving a format and not fallback to any default.

**Comment by litovchenko1** [comment]
VThank you very much for your support
My first function for CMS NEOS.

Examples:
dateObject_1 = ${Date.formatCldr(q(node).property("testMyDate"), 'dd MMMM yyyy HH:mm:ss', 'ru_RU')}
dateObject_2 = ${Date.formatCldr(q(node).property("testMyDate"), 'dd MMMM yyyy HH:mm:ss')}

File: Neos.Eel/Classes/Helper/DateHelper.php [hunk 3]
```diff
@@ -57,6 +71,31 @@ public function format($date, $format)
     }
 
     /**
+     * Format a date to a string with a given cldr format
+     *
+     * @param \DateTime $date
+     * @param string $cldrFormat Format string in CLDR format (see http://cldr.unicode.org/translation/date-time)
+     * @param null|string $locale String locale - example (de|en|ru_RU)
+     *
```
Comment (aertmann) [inline Neos.Eel/Classes/Helper/DateHelper.php]: no empty line between these

**Comment by aertmann** [comment]
regarding documentation, I think that is auto-generated but not sure if it happens regularly for the master branch though

File: Neos.Eel/Classes/Helper/DateHelper.php (hunk 3 above)
Comment (dimaip) [inline Neos.Eel/Classes/Helper/DateHelper.php]: We tried adding the tests, but there are like 5 methods to mock for 3 lines of actual code, don't know if it's worth to load Ivan with such task :)

**Comment by dimaip** [comment]
@kitsunet On the other hand `Date.format` does just that, would make sense to handle input the way it does?

File: Neos.Eel/Classes/Helper/DateHelper.php [hunk 4]
```diff
@@ -13,15 +13,29 @@
 
 use Neos\Flow\Annotations as Flow;
 use Neos\Eel\ProtectedContextAwareInterface;
+use Neos\Flow\I18n\Exception as I18nException;
```
Comment (dimaip) [inline Neos.Eel/Classes/Helper/DateHelper.php]: This exception is currently not used. Remove it if you don't need it.

**Comment by kdambekalns** [comment]
- Eel helper ref is rendered automatically
- why not mock lines for testing 3 lines?
- would be awesome to have this in 4.1, as it seems to be really almost done :)

File: Neos.Eel/Classes/Helper/DateHelper.php [hunk 5]
```diff
@@ -57,6 +71,31 @@ public function format($date, $format)
...
+     * @param \DateTime $date
+     * @param string $cldrFormat Format string in CLDR format (see http://cldr.unicode.org/translation/date-time)
+     * @param null|string $locale String locale - example (de|en|ru_RU)
+     *
+     * @return string
+     */
+    public function formatCldr($date, $cldrFormat, $locale = null)
+    {
+        if (!$date instanceof \DateTimeInterface) {
+            throw new \InvalidArgumentException('"' . $date . '" could not be parsed by \DateTime constructor.');
```
Comment (bwaidelich) [inline Neos.Eel/Classes/Helper/DateHelper.php]: Can't we just use a `\DateTimeInterface` type hint instead?

Comment (bwaidelich) [inline Neos.Eel/Classes/Helper/DateHelper.php]: Also this would break if `$date` is not castable to string

**Comment by dimaip** [comment]
@litovchenko1 how is your progress? Do you need any help to move forward?

**Comment by dimaip** [comment]
@litovchenko1 ping

File: Neos.Eel/Classes/Helper/DateHelper.php (hunk 5 above)
Comment (albe) [inline Neos.Eel/Classes/Helper/DateHelper.php]: good call

File: Neos.Eel/Classes/Helper/DateHelper.php (hunk 3 above)
Comment (albe) [inline Neos.Eel/Classes/Helper/DateHelper.php]: Well, I prefer test coverage tbh. Basically that just shows that this method uses too much delegation, esp. the `$this->localizationService->getConfiguration()->getCurrentLocale();` for example violates the "Law of Demeter". So this should possibly be refactored, either by adding a method to localizationService or by adding a (protected) method in this helper, that fetches the current locale.

**Comment by gerhard-boden** [comment]
@litovchenko1 ping :)

Would like to see this in 4.2, but not too much time left I'm afraid. Any chance you could find the time to get this finished within the upcoming weeks?

**Comment by gerhard-boden** [comment]
Closed in favour of https://github.com/neos/flow-development-collection/pull/1032

**Pull request closed without merging**